lib/utilities/submit_button.dart

```diff
@@ -0,0 +1,52 @@
+import 'package:flutter/material.dart';
+
+class SubmitButton extends StatefulWidget {
+  final Future<void> Function() onPressed;
+  final String label;
+  final bool isButtonDisabled;
+
+  const SubmitButton(
+      {required this.onPressed, required this.label, this.isButtonDisabled = false, Key? key})
+      : super(key: key);
+
+  @override
+  State<SubmitButton> createState() => _SubmitButtonState();
+}
+
+class _SubmitButtonState extends State<SubmitButton> {
+  bool submitting = false;
+
+  @override
+  Widget build(BuildContext context) {
+    return ElevatedButton.icon(
+        label: Text(widget.label),
+        icon: !submitting
+            ? Container()
+            : Container(
+                width: 18,
+                height: 18,
+                padding: const EdgeInsets.all(2.0),
+                child: const CircularProgressIndicator(
+                  color: Colors.white,
+                  strokeWidth: 3,
+                ),
+              ),
+        onPressed: submitting || widget.isButtonDisabled
+            ? null
+            : () async {
+                setState(() {
+                  submitting = true;
+                });
+
+                try {
+                  await widget.onPressed();
+                } on Exception {
+                  rethrow;
+                } finally {
```

Comment:
❓Can you explain this bit?

Reply:
`finally` is executed either way, i.e. it is executed after `try` finishes but it is also executed after `catch` in case of an error.

This is a common exception handling pattern in Java as well - it works pretty much the same in `dart` (there can be tiny differences in behavior when returning from e.g. a finally block, but one should just not do that...).

If it is confusing I can also just do a `try ... catch` and repeat the code of setting `submitting` to false :)

Reply:
I highlighted the wrong bit haha. Sorry!

My question was about why we are `rethrow`ing and what we expect to happen then.